### ISSUE TITLE
`sym` operands can not refer to generic parameters in recent nighty

### ISSUE THREAD

**Issue by yvt** [rust-lang/rust#96330]
I tried this code:

```rust
#![feature(asm_sym)]
#![feature(generic_const_exprs)]

trait Call {
    extern "C" fn call();
}

impl Call for () {
    extern "C" fn call() {}
}

fn call_in_asm<T: Call>() {
    unsafe {
        core::arch::asm!("call {}", sym T::call, clobber_abi("C"));
    }
}

fn main() {
    call_in_asm::<()>();
}
```

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=17566401fba4afe93ec5db938d4fd350))

I expected to see this happen: Successful compilation or compilation failure with an adequate error message

Instead, this happened: Compilation error, `sym` operand marked with `error: unconstrained generic constant`

```
error: unconstrained generic constant
  --> src/main.rs:16:37
   |
16 |         core::arch::asm!("call {}", sym T::call, clobber_abi("C"));
   |                                     ^^^^^^^^^^^
   |
   = help: try adding a `where` bound using this expression: `where [(); sym T::call]:`
```

### Meta

`rustc --version --verbose`:
```
rustc 1.62.0-nightly (de1bc0008 2022-04-21)
binary: rustc
commit-hash: de1bc0008be096cf7ed67b93402250d3b3e480d0
commit-date: 2022-04-21
host: aarch64-apple-darwin
release: 1.62.0-nightly
LLVM version: 14.0.1
```

**Comment by deltragon** [comment]
Did this work before #94468?

**Comment by yvt** [comment]
I think so - this stopped working in `nightly-2022-04-17`, where #94468 was merged.

**Comment by yvt** [comment]
Fixed by #96800
